index.d.ts

```diff
@@ -0,0 +1,205 @@
+  end (result?: string, endTime?: number): void; // TODO: Should we allow number as well for result?
+}
+
+declare class Span extends GenericSpan {
+  // TODO: The following should not be documented right? constructor(), customStackTrace(), setDbContext()
+  transaction: Transaction;
+  name: string | null; // TODO: Should we allow null?
+
+  end (endTime?: number): void;
+}
+
+interface AgentConfigOptions {
+  abortedErrorThreshold?: string | number; // TODO: Do we officially want to support numbers?
+  active?: boolean;
+  apiRequestSize?: string | number; // TODO: Do we officially want to support numbers?
+  apiRequestTime?: string | number; // TODO: Do we officially want to support numbers?
+  asyncHooks?: boolean;
+  captureBody?: 'off' | 'errors' | 'transactions' | 'all';
+  captureErrorLogStackTraces?: 'never' | 'messages' | 'always';
+  captureExceptions?: boolean;
+  captureHeaders?: boolean;
+  captureSpanStackTraces?: boolean;
+  containerId?: string;
+  disableInstrumentations?: string | string[]; // TODO: Do we officially want to support strings?
+  errorMessageMaxLength?: string | number; // TODO: Do we officially want to support numbers?
+  errorOnAbortedRequests?: boolean;
+  filterHttpHeaders?: boolean;
+  frameworkName?: string;
+  frameworkVersion?: string;
+  hostname?: string;
+  ignoreUrls?: Array<string | RegExp>;
+  ignoreUserAgents?: Array<string | RegExp>;
+  instrument?: boolean;
+  kubernetesNamespace?: string;
+  kubernetesNodeName?: string;
+  kubernetesPodName?: string;
+  kubernetesPodUID?: string;
+  logLevel?: 'trace' | 'debug' | 'info' | 'warn' | 'error' | 'fatal';
+  logger?: Logger;
+  metricsInterval?: string | number; // TODO: Do we officially want to support numbers?
+  payloadLogFile?: string;
+  secretToken?: string;
+  serverTimeout?: string | number; // TODO: Do we officially want to support numbers?
+  serverUrl?: string;
+  serviceName?: string;
+  serviceVersion?: string;
+  sourceLinesErrorAppFrames?: number;
+  sourceLinesErrorLibraryFrames?: number;
+  sourceLinesSpanAppFrames?: number;
+  sourceLinesSpanLibraryFrames?: number;
+  stackTraceLimit?: number;
+  transactionMaxSpans?: number;
+  transactionSampleRate?: number;
+  verifyServerCert?: boolean;
+}
+
+interface CaptureErrorOptions {
+  request?: IncomingMessage; // TODO: Currently not documented - should we expose this?
+  response?: ServerResponse; // TODO: Currently not documented - should we expose this?
+  timestamp?: number;
+  handled?: boolean; // TODO: Currently not documented - should we expose this?
+  user?: UserObject;  
+  tags?: Tags; // TODO: Currently not documented
+  custom?: object;
+  message?: string;
+}
+
+interface Tags {
+  [key: string]: TagValue;
+}
+
+interface UserObject {
+  id?: string | number;
+  username?: string;
+  email?: string;
+}
+
+interface ParameterizedMessageObject {
+  message: string;
+  params: Array<any>;
+}
+
+interface Logger {
+  fatal (msg: string, ...args: any[]): void;
```

Comment:
can create an interface for the logger functions
```ts
    interface LoggerFn {
        (msg: string, ...args: any[]): void;
        (obj: object, msg?: string, ...args: any[]): void;
    }
 

   {
     warn: LoggerFn,
     debug: LoggerFn
   }

```

Reply:
Neat

Reply:
I've reverted this optimization again because the auto-complete feature in VS Code wasn't as nice when the methods were defined as properties.

Reply:
I don't see any difference with both changes in auto completion.

Reply:
If you write it as `key: value`, VS Code thinks that this is a property and not a method. Details: https://stackoverflow.com/questions/55208211/how-to-share-a-typescript-function-definition-between-two-classes